plugins/sharding/logical_count.rb

```diff
@@ -56,179 +41,134 @@ def cache_key(input)
-                     line,
-                     method.to_s,
-                     message)
-        end
 
-        def prepare_contexts
-          if @filter or @post_filter
-            if @dynamic_columns.have_initial?
-              apply_targets = []
-              @contexts.each do |context|
-                table = context.table.select_all
-                @temporary_tables << table
-                context.table = table
-                apply_targets << [table]
+          return 0 if @filtered_result_sets.empty?
+
+          total = 0
+          if @window
+            @filtered_result_sets.each do |result_set|
+              @window.each(result_set) do |windowed_result_set|
+                @temporary_tables << windowed_result_set
+                if @context.dynamic_columns.have_filtered?
+                  apply_targets = [[windowed_result_set]]
+                  @context.dynamic_columns.apply_filtered(apply_targets)
+                end
+                unless @post_filter.nil?
+                  windowed_result_set = apply_post_filter(windowed_result_set)
+                  @temporary_tables << windowed_result_set
+                end
+                total += windowed_result_set.size
               end
-              @dynamic_columns.apply_initial(apply_targets)
             end
-          end
-          @contexts.each do |context|
-            filter_shard(context)
-          end
-          if @post_filter
-            if @dynamic_columns.have_filtered?
-              apply_targets = @contexts.collect do |context|
-                [context.table]
+          else
+            apply_filtered_dynamic_columns
+            @filtered_result_sets.each do |result_set|
+              unless @post_filter.nil?
+                result_set = apply_post_filter(result_set)
+                @temporary_tables << result_set
               end
-              @dynamic_columns.apply_filtered(apply_targets)
-            end
-            @contexts.each do |context|
-              post_filter_shard(context)
+              total += result_set.size
             end
           end
+          return total
         end
 
-        def filter_shard(context)
-          return if context.range_index
+        def use_range_index?
+          if @filter or @post_filter
+            decide_use_range_index(false, "need filter",
+                                __LINE__, __method__)
+          elsif @cover_type == :all
+            decide_use_range_index(false, "covered",
+                                __LINE__, __method__)
+          else
+            decide_use_range_index(true, "range index is available",
+                                __LINE__, __method__)
+          end
+        end
 
-          if context.cover_type == :all and @filter.nil?
-            if @post_filter and @dynamic_columns.have_filtered?
-              filtered_table = context.table.select_all
+        def execute_filter(range_index)
+          return if range_index
```

Comment:
Umm... "`execute_filter` does nothing if there is a range index" will confuse developers...
Can we use another approach?

Reply:
I have fixed it. How about the current approach?

- Removed this `return if range_index`.
- Modify to check `@range_index` before filtering in `ShardCountContext.execute`.